Fix logout event sent twice when app is opened the first time

anitaa1990 opened the pull request:
Fixes #1177. When the app is first opened, we fetch the account info (`FETCH_ACCOUNT`)  and the account settings (`FETCH_SETTINGS`) from FluxC. If the user is not logged in, both of these actions return separate errors (`ACCOUNT_FETCH_ERROR` and `SETTINGS_FETCH_GENERIC_ERROR`). Since we do not check the error type before sending the logout event was sent twice.

### Changes
I first thought of sending the `account_logout` event when the error type = `ACCOUNT_FETCH_ERROR`. Something like this: `event.error.type == ACCOUNT_FETCH_ERROR`. 
But if in the future, the error type is null, then the events would not be sent. So this PR adds a new variable called `accountLogoutEventRegistered` which is `false` by default. If the `account_logout` event is already sent, this flag would be set to `true`.

### Behaviour before the fix:
<img src="https://user-images.githubusercontent.com/22608780/60080823-3ee5c680-974e-11e9-97a7-e4c0725ff939.gif">

### Behaviour after the fix:
<img src="https://user-images.githubusercontent.com/22608780/60165443-e9c4b600-981c-11e9-85e6-3b2699d53dea.gif">

anitaa1990 commented:
> hi @anitaa1990 I tested the fix and it does prevent duplicate logout events on first launch, but if the user logs out of the device during the same session then the logout event is not recorded.

Thanks for the review @AmandaRiu and the really good catch! I should have caught this myself 🤦‍♀. 

So the way I understand it, when a user logs out of the app, we sent a logout event, reset their zendesk credentials and delete all user specific values from the local cache. But if a user is not logged in, and the app is opened at the beginning, the `account_logout` event should not be called. Is my understanding correct? It should only be called if the user is clicking on logout button and has been successfully logged out?

#### Current Implementation
The `onAccountChanged` method is called whenever there is a change in the account info for the user. When the app is first opened, we fetch the account info (`FETCH_ACCOUNT`) and the account settings (`FETCH_SETTINGS`) from FluxC. And if the user is not logged in, both of these actions return separate errors (`ACCOUNT_FETCH_ERROR` and `SETTINGS_FETCH_GENERIC_ERROR`). At the moment, for both these errors, the logout related events (such as resetting Zendesk or sending logout event) are being called. 

#### Expected Behaviour and fix
To fix this scenario and to process logout related events only after a user logs out, I was thinking we could check the `event.causeOfChange` and `event.error`. Something like this:
```
    fun onAccountChanged(event: OnAccountChanged) {
      // check if the event is a logout event and only then process logout related tasks
        val isLoggedOut = event.causeOfChange == null && event.error == null
        if (!accountStore.hasAccessToken() && isLoggedOut) {
            // Logged out
            AnalyticsTracker.track(Stat.ACCOUNT_LOGOUT)

            // Reset analytics
            AnalyticsTracker.flush()
            AnalyticsTracker.clearAllData()
            CrashUtils.resetAccountAndSite()
            zendeskHelper.reset()

            // Wipe user-specific preferences
            AppPrefs.reset()
        }
    }
```

This would only be true if the user is successfully logged out of the app and would not be called when the app is opened for the first time. Does this make sense?

AmandaRiu commented:
> This would only be true if the user is successfully logged out of the app and would not be called when the app is opened for the first time. Does this make sense?

Absolutely :) I agree we should only log the `ACCOUNT_LOGOUT` event if the user is actually logging out. It's strange that the FluxC event is not sending the `causeOfChange` for `AccountAction.SIGN_OUT`, being able to check for that would be a really solid fix, but I'm not sure if adding it to the FluxC event would somehow break the way WordPress-Android uses it so that would take a little more research. For now, the fix you outlined above looks like it should work.

anitaa1990 commented:
Thanks for the clarification @AmandaRiu! I've reverted the previous logic and added the modifications to the PR in 68595f4. Ready for another round of review 👍
